net: avoid extra dns query per seed

theuni opened the pull request:
This addresses the TODO here: https://github.com/bitcoin/bitcoin/blob/master/src/net.cpp#L1604 and is necessary before introducing async dns queries.

Right now, we resolve an extra hostname for each dns seed, so that we have some source ip to identify them with in addrman. It's entirely arbitrary and unnecessary.

Instead, create a dummy "ip" for this association with a prefix of fd6b:88c0:8724::/48 (fd + sha256(bitcoin)[0:5]).

I've also changed the "source ip" to match a hash of the host with the subdomain included, rather than having all results mapped to the top-level domain.

I'm not sure whether we need to bother with addrman forwards-compatibility for graceful downgrades. Thoughts on that?

theuni committed:
net: add an internal subnet for representing unresolved hostnames

We currently do two resolves for dns seeds: one for the results, and one to
serve in addrman as the source for those addresses.

There's no requirement that the source hostname resolves to the stored
identifier, only that the mapping is unique. So rather than incurring the
second lookup, combine a private subnet with a hash of the hostname.

The resulting v6 ip is guaranteed not to be publicy routable, and has only a
negligible chance of colliding with a user's internal network (which would be
of no consequence anyway).

theuni committed:
net: switch to dummy internal ip for dns seed source

This addresss the TODO to avoid resolving twice.

laanwj commented:
Concept ACK

gmaxwell commented:
Concept ACK.

TheBlueMatt commented:
It would be kinda nice if we blocked DNS names from resolving to internal IPs, since that should never happen, and unlike local/tor/whatever, there is no reason you'd ever want to -connect to them.

theuni commented:
@TheBlueMatt Agreed in principle. Though they do fail the IsValid test, so we'd never actually attempt a connection to them: https://github.com/bitcoin/bitcoin/pull/10446/files#diff-b64569708508232923e5fe3059396334R216

I can add explicit filtering to resolving if that would make you more comfortable?

sipa commented:
Concept ACK, but can you adapt the touched code to the style guide?

Agree with @TheBlueMatt, but maybe we should also do that for onion addresses?

TheBlueMatt commented:
@sipa well dns seeds and other dns queries are allowed to return bitcoin-onion-encoded results (I've done this as -addnode/-connect arguments in the past).

@theuni Yea, I would feel a bit more comfortable doing that, thanks :).

theuni committed:
net: do not allow resolving to an internal address

In order to prevent mixups, our internal range is never allowed as a resolve
result. This means that no user-provided string will ever be confused with an
internal address.

theuni commented:
Since there have been no (ut)ACKs and there were several things to change, I went ahead and made changes, squashed, and rebased. I believe this is good to go now.

changes since the first round:
- rebased/squashed
- adapted code style
- use sha256 rather than sha256d for name mapping, for easier sanity checking
- filter out internal addresses from the resolver
- minor fixes

theuni commented:
@sipa by filtering onion addresses the same way that internal ones are done here, we'd prevent dns seeds from handing out onion addresses. I'm not sure if the current seeders include those (I think they do), but I see no reason to filter them either way.

sipa commented on src/netaddress.h:
```diff
@@ -45,6 +46,12 @@ class CNetAddr
 
+        /**
+          * Transform an arbitrary string into a non-routable ipv6 address.
+          * Useful for mapping resolved addresses back to thir source.
```
VERY BIG ERROR WOW: 'thir'

theuni replied:
What, no responsible disclosure? Jeez!

sipa reviewed (commented):
utACK

theuni committed:
net: use an internal address for fixed seeds

laanwj commented on src/chainparams.cpp:
```diff
@@ -124,12 +124,12 @@ class CMainParams : public CChainParams {
-        vSeeds.push_back(CDNSSeedData("petertodd.org", "seed.btc.petertodd.org", true)); // Peter Todd, only supports x1, x5, x9, and xd
+        vSeeds.emplace_back("seed.bitcoin.sipa.be", true); // Pieter Wuille, only supports x1, x5, x9, and xd
+        vSeeds.emplace_back("dnsseed.bluematt.me", true); // Matt Corallo, only supports x9
+        vSeeds.emplace_back("dnsseed.bitcoin.dashjr.org"); // Luke Dashjr
```
Sorry for the last minute comment, feel free to say you don't want to do it in this PR. But as all these lines are updated I think this is  a nice opportunity to remove the default argument for `supportsServiceBitsFilteringIn` and add explicit `, false` here.

theuni replied:
No problem, I found one tiny additional thing to change on this today, I'll go ahead with your suggestion while I'm at it.

theuni replied:
Pushed a small commit for this to avoid invalidating review.

theuni committed:
chainparams: make supported service bits option explicit

theuni commented:
Added a very small commit to use an internal address for seed nodes. I believe this is ready for merge.

TheBlueMatt commented:
utACK c1be28536467e90ce75eaa7d8c338f6485c4bee5

laanwj commented:
utACK c1be285
